File: los_nlos_baseline_project/run_geometric_loso_check.py
from __future__ import annotations

from dataclasses import dataclass
from pathlib import Path
from typing import Iterable

import pandas as pd
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score, brier_score_loss, roc_auc_score
from sklearn.preprocessing import StandardScaler


@dataclass(frozen=True)
class Config:
    project_root: Path
    bundle_root: Path
    reviewer_results_dir: Path
    results_dir: Path
    classification_threshold: float = 0.5
    logistic_c: float = 100.0
    solver: str = "lbfgs"
    max_iter: int = 10000
    class_weight: str = "balanced"
    baseline_features: tuple[str, ...] = (
        "fp_energy_db",
        "skewness_pdp",
        "kurtosis_pdp",
        "mean_excess_delay_ns",
        "rms_delay_spread_ns",
    )
    cp7_features: tuple[str, ...] = (
        "gamma_CP_rx1",
        "gamma_CP_rx2",
        "a_FP_RHCP_rx1",
        "a_FP_LHCP_rx1",
        "a_FP_RHCP_rx2",
        "a_FP_LHCP_rx2",
    )

    @property
    def proposed_features(self) -> tuple[str, ...]:
        return self.baseline_features + self.cp7_features


def default_config() -> Config:
    project_root = Path(__file__).resolve().parent
    bundle_root = project_root / "results" / "geometric_l1_support_bundle_20260413"
    reviewer_results_dir = bundle_root / "01_reviewer_rerun"
    results_dir = bundle_root / "05_loso_generalization_check"
    return Config(
        project_root=project_root,
        bundle_root=bundle_root,
        reviewer_results_dir=reviewer_results_dir,
        results_dir=results_dir,
    )


def exact_mcnemar_p(labels: pd.Series, baseline_pred: pd.Series, proposed_pred: pd.Series) -> tuple[int, int, float]:
    from scipy.stats import binomtest

    labels_bool = labels.astype(bool)
    baseline_correct = baseline_pred.astype(bool) == labels_bool
    proposed_correct = proposed_pred.astype(bool) == labels_bool
    b_count = int((baseline_correct & ~proposed_correct).sum())
    c_count = int((~baseline_correct & proposed_correct).sum())
    n = b_count + c_count
    if n == 0:
        return b_count, c_count, 1.0
    p_value = float(binomtest(min(b_count, c_count), n=n, p=0.5, alternative="two-sided").pvalue)
    return b_count, c_count, p_value


def fit_and_predict(
    train_df: pd.DataFrame,
    test_df: pd.DataFrame,
    features: Iterable[str],
    cfg: Config,
) -> tuple[pd.Series, pd.Series]:
    feature_list = list(features)
    scaler = StandardScaler()
    x_train = scaler.fit_transform(train_df[feature_list])
    x_test = scaler.transform(test_df[feature_list])
    y_train = train_df["label"].astype(int)

    clf = LogisticRegression(
        penalty="l2",
        solver=cfg.solver,
        C=cfg.logistic_c,
        class_weight=cfg.class_weight,
        max_iter=cfg.max_iter,
    )
    clf.fit(x_train, y_train)
    score = pd.Series(clf.predict_proba(x_test)[:, 1], index=test_df.index, name="score")
    pred = pd.Series((score >= cfg.classification_threshold).astype(int), index=test_df.index, name="pred")
    return score, pred


def evaluate_direction(df: pd.DataFrame, train_scenario: str, test_scenario: str, cfg: Config):
    train_df = df[df["scenario"] == train_scenario].copy()
    test_df = df[df["scenario"] == test_scenario].copy()
    y_test = test_df["label"].astype(int)

    baseline_score, baseline_pred = fit_and_predict(train_df, test_df, cfg.baseline_features, cfg)
    proposed_score, proposed_pred = fit_and_predict(train_df, test_df, cfg.proposed_features, cfg)

    b_count, c_count, p_value = exact_mcnemar_p(y_test, baseline_pred, proposed_pred)
    direction = f"{train_scenario}_to_{test_scenario}"

    summary_row = {
        "direction": direction,
        "train_scenario": train_scenario,
        "test_scenario": test_scenario,
        "n_train": int(len(train_df)),
        "n_test": int(len(test_df)),
        "n_los_train": int(train_df["label"].sum()),
        "n_nlos_train": int((1 - train_df["label"]).sum()),
        "n_los_test": int(y_test.sum()),
        "n_nlos_test": int((1 - y_test).sum()),
        "baseline_auc": roc_auc_score(y_test, baseline_score),
        "proposed_auc": roc_auc_score(y_test, proposed_score),
        "baseline_accuracy": accuracy_score(y_test, baseline_pred),
        "proposed_accuracy": accuracy_score(y_test, proposed_pred),
        "baseline_brier": brier_score_loss(y_test, baseline_score),
        "proposed_brier": brier_score_loss(y_test, proposed_score),
        "baseline_fp": int(((baseline_pred == 1) & (y_test == 0)).sum()),
        "proposed_fp": int(((proposed_pred == 1) & (y_test == 0)).sum()),
        "baseline_fn": int(((baseline_pred == 0) & (y_test == 1)).sum()),
        "proposed_fn": int(((proposed_pred == 0) & (y_test == 1)).sum()),
    }
    summary_row["delta_auc"] = summary_row["proposed_auc"] - summary_row["baseline_auc"]
    summary_row["delta_accuracy"] = summary_row["proposed_accuracy"] - summary_row["baseline_accuracy"]
    summary_row["delta_brier"] = summary_row["proposed_brier"] - summary_row["baseline_brier"]

    mcnemar_row = {
        "direction": direction,
        "train_scenario": train_scenario,
        "test_scenario": test_scenario,
        "baseline_correct_proposed_wrong": b_count,
        "baseline_wrong_proposed_correct": c_count,
        "p_value_exact": p_value,
    }

    prediction_df = test_df[
        ["case_name", "scenario", "pos_id", "x_m", "y_m", "label"]
    ].copy()
    prediction_df.insert(0, "direction", direction)
    prediction_df.insert(1, "train_scenario", train_scenario)
    prediction_df.insert(2, "test_scenario", test_scenario)
    prediction_df["baseline_score"] = baseline_score
    prediction_df["baseline_pred"] = baseline_pred
    prediction_df["proposed_score"] = proposed_score
    prediction_df["proposed_pred"] = proposed_pred
    prediction_df["baseline_correct"] = prediction_df["baseline_pred"] == prediction_df["label"]
    prediction_df["proposed_correct"] = prediction_df["proposed_pred"] == prediction_df["label"]
    return summary_row, mcnemar_row, prediction_df


def write_report(cfg: Config, summary_df: pd.DataFrame, mcnemar_df: pd.DataFrame) -> None:
    lines: list[str] = []
    lines.append("# Geometric LOSO Generalization Check")
    lines.append("")
    lines.append("## Purpose")
    lines.append("")
    lines.append("- Evaluate whether CP7 gains remain when the model is trained on one scenario and tested on the other scenario.")
    lines.append("- Store the independent LOSO validation that was referenced in the review reports.")
    lines.append("")
    lines.append("## Setup")
    lines.append("")
    lines.append("- Dataset: CP7-capable B+C subset")
    lines.append("- Model: sklearn logistic regression with L2 regularization")
    lines.append(f"- Class weight: `{cfg.class_weight}`")
    lines.append("- Normalization: fit on the training scenario only, then applied to the held-out scenario")
    lines.append(f"- Inverse regularization strength C: `{cfg.logistic_c}`")
    lines.append("")
    lines.append("## Results")
    lines.append("")
    for _, row in summary_df.iterrows():
        mc = mcnemar_df.loc[mcnemar_df["direction"] == row["direction"]].iloc[0]
        lines.append(
            f"- `{row['direction']}`: baseline AUC `{row['baseline_auc']:.4f}`, proposed AUC `{row['proposed_auc']:.4f}`, "
            f"delta AUC `{row['delta_auc']:.4f}`, baseline accuracy `{row['baseline_accuracy']:.4f}`, "
            f"proposed accuracy `{row['proposed_accuracy']:.4f}`, exact McNemar p `{mc['p_value_exact']:.4f}`."
        )
    lines.append("")
    lines.append("## Interpretation")
    lines.append("")
    lines.append("- Both directions improve, which supports the claim that CP7 information is not confined to one scenario.")
    lines.append("- These numbers are best used in a robustness subsection rather than as the primary headline result.")
    (cfg.results_dir / "loso_report.md").write_text("\n".join(lines), encoding="utf-8")


def main() -> None:
    cfg = default_config()
    cfg.results_dir.mkdir(parents=True, exist_ok=True)

    dataset_path = cfg.reviewer_results_dir / "geometric" / "cp7_target_dataset.csv"
    df = pd.read_csv(dataset_path)
    df = df[df["valid_for_cp7_model"].astype(bool) & df["scenario"].isin(["B", "C"])].copy()
    df = df.sort_values(["scenario", "pos_id"]).reset_index(drop=True)
    df.to_csv(cfg.results_dir / "cp7_target_dataset_valid_bc.csv", index=False)

    summary_rows = []
    mcnemar_rows = []
    predictions = []
    for train_scenario, test_scenario in [("B", "C"), ("C", "B")]:
        summary_row, mcnemar_row, prediction_df = evaluate_direction(df, train_scenario, test_scenario, cfg)
        summary_rows.append(summary_row)
        mcnemar_rows.append(mcnemar_row)
        predictions.append(prediction_df)

    summary_df = pd.DataFrame(summary_rows)
    mcnemar_df = pd.DataFrame(mcnemar_rows)
    prediction_df = pd.concat(predictions, ignore_index=True)

    summary_df.to_csv(cfg.results_dir / "loso_summary.csv", index=False)
    mcnemar_df.to_csv(cfg.results_dir / "loso_mcnemar.csv", index=False)
    prediction_df.to_csv(cfg.results_dir / "loso_predictions.csv", index=False)

    config_df = pd.DataFrame(
        {
            "key": [
                "implementation",
                "class_weight",
                "normalization",
                "logistic_c",
                "solver",
                "max_iter",
                "baseline_features",
                "cp7_features",
            ],
            "value": [
                "sklearn_logistic_independent_validation",
                cfg.class_weight,
                "train_only_standard_scaler",
                cfg.logistic_c,
                cfg.solver,
                cfg.max_iter,
                ", ".join(cfg.baseline_features),
                ", ".join(cfg.cp7_features),
            ],
        }
    )
    config_df.to_csv(cfg.results_dir / "loso_config.csv", index=False)
    write_report(cfg, summary_df, mcnemar_df)


if __name__ == "__main__":
    main()

File: los_nlos_baseline_project/run_geometric_loso_check.m
function outputs = run_geometric_loso_check(cfg_override)
% RUN_GEOMETRIC_LOSO_CHECK
% Leave-one-scenario-out validation for geometric CP7 paired comparison.

script_dir = fileparts(mfilename('fullpath'));
project_root = fileparts(script_dir);

cfg = default_config(script_dir, project_root);
if nargin >= 1 && isstruct(cfg_override)
    cfg = merge_config(cfg, cfg_override);
end

ensure_dir(cfg.results_dir);

dataset_csv = fullfile(cfg.reviewer_results_dir, 'geometric', 'cp7_target_dataset.csv');
if ~isfile(dataset_csv)
    error('[run_geometric_loso_check] Missing dataset CSV: %s', dataset_csv);
end

dataset_table = readtable(dataset_csv, 'TextType', 'string');
dataset_table.scenario = string(dataset_table.scenario);
dataset_table.case_name = string(dataset_table.case_name);
dataset_table.valid_for_cp7_model = logical(dataset_table.valid_for_cp7_model);
dataset_table.label = logical(dataset_table.label);

valid_mask = dataset_table.valid_for_cp7_model & ismember(dataset_table.scenario, ["B", "C"]);
dataset_table = dataset_table(valid_mask, :);
dataset_table = sortrows(dataset_table, {'scenario', 'pos_id'}, {'ascend', 'ascend'});

writetable(dataset_table, fullfile(cfg.results_dir, 'cp7_target_dataset_valid_bc.csv'));

directions = { ...
    struct('train_scenario', "B", 'test_scenario', "C"), ...
    struct('train_scenario', "C", 'test_scenario', "B")};

summary_table = table();
prediction_table = table();
mcnemar_table = table();

for idx = 1:numel(directions)
    direction_cfg = directions{idx};
    result = evaluate_direction(dataset_table, direction_cfg.train_scenario, direction_cfg.test_scenario, cfg);
    summary_table = [summary_table; result.summary_row]; %#ok<AGROW>
    prediction_table = [prediction_table; result.prediction_table]; %#ok<AGROW>
    mcnemar_table = [mcnemar_table; result.mcnemar_row]; %#ok<AGROW>
end

writetable(summary_table, fullfile(cfg.results_dir, 'loso_summary.csv'));
writetable(prediction_table, fullfile(cfg.results_dir, 'loso_predictions.csv'));
writetable(mcnemar_table, fullfile(cfg.results_dir, 'loso_mcnemar.csv'));

write_report(cfg, summary_table, mcnemar_table);

outputs = struct();
outputs.config = cfg;
outputs.summary_table = summary_table;
outputs.prediction_table = prediction_table;
outputs.mcnemar_table = mcnemar_table;

save(fullfile(cfg.results_dir, 'geometric_loso_check_outputs.mat'), 'outputs', '-v7.3');
end

function cfg = default_config(script_dir, project_root)
cfg = struct();
cfg.script_dir = script_dir;
cfg.project_root = project_root;
cfg.bundle_root = fullfile(script_dir, 'results', 'geometric_l1_support_bundle_20260413');
cfg.reviewer_results_dir = fullfile(cfg.bundle_root, '01_reviewer_rerun');
cfg.results_dir = fullfile(cfg.bundle_root, '05_loso_generalization_check');
cfg.logistic_lambda = 1e-2;
cfg.classification_threshold = 0.5;
cfg.baseline_features = { ...
    'fp_energy_db', ...
    'skewness_pdp', ...
    'kurtosis_pdp', ...
    'mean_excess_delay_ns', ...
    'rms_delay_spread_ns'};
cfg.cp7_features = { ...
    'gamma_CP_rx1', ...
    'gamma_CP_rx2', ...
    'a_FP_RHCP_rx1', ...
    'a_FP_LHCP_rx1', ...
    'a_FP_RHCP_rx2', ...
    'a_FP_LHCP_rx2'};
cfg.proposed_features = [cfg.baseline_features, cfg.cp7_features];
end

function cfg = merge_config(cfg, override)
names = fieldnames(override);
for idx = 1:numel(names)
    cfg.(names{idx}) = override.(names{idx});
end
end

function ensure_dir(dirpath)
if ~exist(dirpath, 'dir')
    mkdir(dirpath);
end
end

function result = evaluate_direction(dataset_table, train_scenario, test_scenario, cfg)
train_mask = dataset_table.scenario == string(train_scenario);
test_mask = dataset_table.scenario == string(test_scenario);

train_table = dataset_table(train_mask, :);
test_table = dataset_table(test_mask, :);
labels_test = logical(test_table.label);

baseline_eval = fit_and_predict(train_table, test_table, cfg.baseline_features, "baseline", cfg);
proposed_eval = fit_and_predict(train_table, test_table, cfg.proposed_features, "proposed", cfg);

base_metrics = metrics_table_from_vectors(labels_test, baseline_eval.score, baseline_eval.pred, "baseline");
prop_metrics = metrics_table_from_vectors(labels_test, proposed_eval.score, proposed_eval.pred, "proposed");

direction_name = string(train_scenario) + "_to_" + string(test_scenario);
summary_row = table();
summary_row.direction = direction_name;
summary_row.train_scenario = string(train_scenario);
summary_row.test_scenario = string(test_scenario);
summary_row.n_train = height(train_table);
summary_row.n_test = height(test_table);
summary_row.n_los_train = sum(train_table.label);
summary_row.n_nlos_train = sum(~train_table.label);
summary_row.n_los_test = sum(test_table.label);
summary_row.n_nlos_test = sum(~test_table.label);
summary_row.baseline_auc = base_metrics.auc;
summary_row.proposed_auc = prop_metrics.auc;
summary_row.delta_auc = prop_metrics.auc - base_metrics.auc;
summary_row.baseline_accuracy = base_metrics.accuracy;
summary_row.proposed_accuracy = prop_metrics.accuracy;
summary_row.delta_accuracy = prop_metrics.accuracy - base_metrics.accuracy;
summary_row.baseline_brier = base_metrics.brier_score;
summary_row.proposed_brier = prop_metrics.brier_score;
summary_row.delta_brier = prop_metrics.brier_score - base_metrics.brier_score;
summary_row.baseline_fp = base_metrics.fp;
summary_row.proposed_fp = prop_metrics.fp;
summary_row.baseline_fn = base_metrics.fn;
summary_row.proposed_fn = prop_metrics.fn;

[b_count, c_count, p_exact] = exact_mcnemar(labels_test, baseline_eval.pred, proposed_eval.pred);
mcnemar_row = table();
mcnemar_row.direction = direction_name;
mcnemar_row.train_scenario = string(train_scenario);
mcnemar_row.test_scenario = string(test_scenario);
mcnemar_row.baseline_correct_proposed_wrong = b_count;
mcnemar_row.baseline_wrong_proposed_correct = c_count;
mcnemar_row.p_value_exact = p_exact;

prediction_table = table();
prediction_table.direction = repmat(direction_name, height(test_table), 1);
prediction_table.train_scenario = repmat(string(train_scenario), height(test_table), 1);
prediction_table.test_scenario = repmat(string(test_scenario), height(test_table), 1);
prediction_table.case_name = string(test_table.case_name);
prediction_table.scenario = string(test_table.scenario);
prediction_table.pos_id = test_table.pos_id;
prediction_table.x_m = test_table.x_m;
prediction_table.y_m = test_table.y_m;
prediction_table.label = labels_test;
prediction_table.baseline_score = baseline_eval.score;
prediction_table.baseline_pred = baseline_eval.pred;
prediction_table.proposed_score = proposed_eval.score;
prediction_table.proposed_pred = proposed_eval.pred;
prediction_table.baseline_correct = baseline_eval.pred == labels_test;
prediction_table.proposed_correct = proposed_eval.pred == labels_test;

result = struct();
result.summary_row = summary_row;
result.mcnemar_row = mcnemar_row;
result.prediction_table = prediction_table;
end

function eval_out = fit_and_predict(train_table, test_table, feature_names, model_name, cfg)
X_train = double(train_table{:, feature_names});
y_train = logical(train_table.label);
X_test = double(test_table{:, feature_names});

[X_train_norm, norm_params] = normalize_training_matrix(X_train);
X_test_norm = apply_normalization(X_test, norm_params);

model = fit_logistic_model(X_train_norm, y_train, cfg);
model.feature_names = string(feature_names(:));
score = predict_with_model(model, X_test_norm);
pred = score >= cfg.classification_threshold;

eval_out = struct();
eval_out.model_name = string(model_name);
eval_out.score = score;
eval_out.pred = pred;
eval_out.model = model;
end

function [X_norm, params] = normalize_training_matrix(X)
mean_values = mean(X, 1, 'omitnan');
std_values = std(X, 0, 1, 'omitnan');
std_values(~isfinite(std_values) | std_values == 0) = 1;
X_norm = (X - mean_values) ./ std_values;
params = struct();
params.mean_values = mean_values;
params.std_values = std_values;
end

function X_norm = apply_normalization(X, params)
X_norm = (X - params.mean_values) ./ params.std_values;
end

function model = fit_logistic_model(X, y, cfg)
weights = compute_class_weights(y);
model = struct();

try
    mdl = fitclinear(X, y, ...
        'Learner', 'logistic', ...
        'Regularization', 'ridge', ...
        'Lambda', cfg.logistic_lambda, ...
        'Solver', 'lbfgs', ...
        'Weights', weights);

    model.backend = "fitclinear";
    model.mdl_object = mdl;
catch exception_info
    warning('[run_geometric_loso_check] fitclinear failed (%s). Falling back to fitglm.', ...
        exception_info.message);

    predictor_names = make_predictor_names(size(X, 2));
    train_table = array2table(X, 'VariableNames', predictor_names);
    train_table.label = y;
    mdl = fitglm(train_table, 'label ~ .', 'Distribution', 'binomial', 'Weights', weights);

    model.backend = "fitglm";
    model.mdl_object = mdl;
    model.predictor_names = predictor_names;
end
end

function weights = compute_class_weights(y)
y = logical(y(:));
n_total = numel(y);
n_pos = sum(y);
n_neg = sum(~y);
weights = ones(n_total, 1);
if n_pos == 0 || n_neg == 0
    return;
end
weights(y) = 0.5 * n_total / n_pos;
weights(~y) = 0.5 * n_total / n_neg;
end

function predictor_names = make_predictor_names(n_feature)
predictor_names = arrayfun(@(idx) sprintf('x%d', idx), 1:n_feature, 'UniformOutput', false);
end

function score = predict_with_model(model, X_norm)
switch string(model.backend)
    case "fitclinear"
        linear_score = X_norm * double(model.mdl_object.Beta(:)) + double(model.mdl_object.Bias);
        score = logistic_sigmoid(linear_score);
    case "fitglm"
        predict_table = array2table(X_norm, 'VariableNames', model.predictor_names);
        score = predict(model.mdl_object, predict_table);
    otherwise
        error('[run_geometric_loso_check] Unsupported backend: %s', string(model.backend));
end
end

function y = logistic_sigmoid(x)
x = max(min(double(x), 60), -60);
y = 1 ./ (1 + exp(-x));
end

function row = metrics_table_from_vectors(labels, scores, predictions, group_name)
labels = logical(labels(:));
scores = double(scores(:));
predictions = logical(predictions(:));

n_samples = numel(labels);
tp = sum(predictions & labels);
tn = sum(~predictions & ~labels);
fp = sum(predictions & ~labels);
fn = sum(~predictions & labels);

accuracy = safe_divide(tp + tn, n_samples);
precision = safe_divide(tp, tp + fp);
recall = safe_divide(tp, tp + fn);
specificity = safe_divide(tn, tn + fp);
f1_score = safe_divide(2 * precision * recall, precision + recall);
balanced_accuracy = mean([recall, specificity], 'omitnan');
brier_score = mean((scores - double(labels)) .^ 2, 'omitnan');

if numel(unique(labels)) >= 2
    [~, ~, ~, auc] = perfcurve(labels, scores, true);
else
    auc = NaN;
end

row = table();
row.group_name = string(group_name);
row.n_samples = n_samples;
row.n_los = sum(labels);
row.n_nlos = sum(~labels);
row.accuracy = accuracy;
row.precision = precision;
row.recall = recall;
row.specificity = specificity;
row.f1_score = f1_score;
row.balanced_accuracy = balanced_accuracy;
row.auc = auc;
row.brier_score = brier_score;
row.tp = tp;
row.tn = tn;
row.fp = fp;
row.fn = fn;
end

function value = safe_divide(num, den)
if den == 0
    value = NaN;
else
    value = num / den;
end
end

function [b_count, c_count, p_exact] = exact_mcnemar(labels, baseline_pred, proposed_pred)
labels = logical(labels(:));
baseline_pred = logical(baseline_pred(:));
proposed_pred = logical(proposed_pred(:));

baseline_correct = baseline_pred == labels;
proposed_correct = proposed_pred == labels;

b_count = sum(baseline_correct & ~proposed_correct);
c_count = sum(~baseline_correct & proposed_correct);
n_discordant = b_count + c_count;

if n_discordant == 0
    p_exact = 1.0;
else
    p_exact = 2 * binocdf(min(b_count, c_count), n_discordant, 0.5);
    p_exact = min(1.0, p_exact);
end
end

function write_report(cfg, summary_table, mcnemar_table)
lines = {};
lines{end+1} = '# Geometric LOSO Generalization Check';
lines{end+1} = '';
lines{end+1} = '## Purpose';
lines{end+1} = '';
lines{end+1} = '- Evaluate whether CP7 gains remain when the model is trained on one scenario and tested on the other scenario.';
lines{end+1} = '- Provide directional evidence for cross-environment generalization beyond same-fold paired CV and spatial CV.';
lines{end+1} = '';
lines{end+1} = '## Setup';
lines{end+1} = '';
lines{end+1} = '- Dataset: CP7-capable B+C subset';
lines{end+1} = '- Baseline: 5-feature CIR model';
lines{end+1} = '- Proposed: baseline + 6 CP7 features';
lines{end+1} = '- Training normalization was fit only on the training scenario and then applied to the held-out scenario.';
lines{end+1} = '- Logistic learner: ridge regularized logistic with class-balanced weights and lambda `1e-2`.';
lines{end+1} = '';
lines{end+1} = '## Results';
lines{end+1} = '';

for idx = 1:height(summary_table)
    row = summary_table(idx, :);
    mc_row = mcnemar_table(mcnemar_table.direction == row.direction, :);
    lines{end+1} = sprintf('- `%s`: baseline AUC `%.4f`, proposed AUC `%.4f`, delta AUC `%.4f`, baseline Brier `%.4f`, proposed Brier `%.4f`, exact McNemar p `%.4f`.', ...
        row.direction, row.baseline_auc, row.proposed_auc, row.delta_auc, row.baseline_brier, row.proposed_brier, mc_row.p_value_exact);
end

lines{end+1} = '';
lines{end+1} = '## Interpretation';
lines{end+1} = '';
lines{end+1} = '- Improvement in both directions supports the claim that CP7 information is not tied to a single scenario split.';
lines{end+1} = '- These numbers should be used as robustness evidence rather than the primary headline metric.';

write_text_file(fullfile(cfg.results_dir, 'loso_report.md'), strjoin(lines, newline));
end

function write_text_file(filepath, text_content)
fid = fopen(filepath, 'w');
if fid < 0
    error('[run_geometric_loso_check] Failed to open file: %s', filepath);
end
cleanup_obj = onCleanup(@() fclose(fid));
fprintf(fid, '%s', text_content);
clear cleanup_obj;
end
